### PR TITLE
Basic Scaffolding

### LINKED ISSUE #2: [Content Publishing] Add Basic Scaffolding/ Boiletplate Nestjs

### PR DESCRIPTION
**Pull request by saraswatpuneet**
* Github actions - build, test and release
* Nestjs boilerplate with `Blockchain support`
* Main service controller/module/class
* etc.

Closes #2

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      - name: Install Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: 18
+          cache: 'npm'
+          registry-url: 'https://registry.npmjs.org'
+          cache-dependency-path: package-lock.json
+      - name: Install dependencies
+        run: npm ci
+      - name: Build Next.js
```
Comment (aramikm) [inline .github/workflows/build.yml:27]: nit: next or nest?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        with:
+          node-version: 18
+          cache: 'npm'
+          registry-url: 'https://registry.npmjs.org'
+          cache-dependency-path: package-lock.json
+      - name: Install dependencies
+        run: npm ci
+      - name: License Check
+        # List all the licenses and error out if it is not one of the supported licenses
+        run: npx license-report --fields=name --fields=licenseType | jq 'map(select(.licenseType | IN("MIT", "Apache-2.0", "ISC", "BSD-3-Clause", "BSD-2-Clause") | not)) | if length == 0 then halt else halt_error(1) end'
```
Comment (aramikm) [inline .github/workflows/build.yml:59]: this is great, I wish we had something like that for graph-sdk

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      - 'v0.0.1-rc[0-9]+' # used for testing only
+  workflow_dispatch:
+    inputs:
+      release-version:
+        description: 'Release version (v#.#.#[-rc#])'
+        required: true
+
+env:
+  NEW_RELEASE_TAG_FROM_UI: ${{github.event.inputs.release-version}}
+  TEST_RUN: ${{startsWith(github.event.inputs.release-version || github.ref_name, 'v0.0.1')}}
```
Comment (aramikm) [inline .github/workflows/release.yml:21]: do we publish packages on `rc` releases?

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+RUN apt-get -y install redis
+RUN sed -e 's/^appendonly .*$/appendonly yes/' /etc/redis/redis.conf > /etc/redis/redis.conf.appendonly
+RUN mv /etc/redis/redis.conf.appendonly /etc/redis/redis.conf
+
+ENV REDIS_URL=redis://localhost:6379
+
+VOLUME [ "/var/lib/redis" ]
+
+# Start the application
+ENTRYPOINT service redis-server start && npm start
```
Comment (aramikm) [inline Dockerfile:32]: there is an assumption that we might use redis which might be correct but other than that shouldn't Redis start before the app?

File: INSTALLING.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+For the application to start & run correctly, it is necessary to configure the environment with certain parameters. These should be injected into a container pod if running in a containerized environment.
+
+The following is a list of environment variables that may be set to control the application's behavior and environment. The complete list can always be referenced [here](./env.template)
+
+|Variable|required?|Description|Default|
+|-|-|-|-|
+|`FREQUENCY_URL`|**yes**|Blockchain URL|_none_|
+|`PROVIDER_ID`|**yes**|MSA ID of provider|_none_|
+|`PROVIDER_BASE_URL`|**yes**|URL of provider graph query endpoint|_none_|
```
Comment (aramikm) [inline INSTALLING.md:65]: nit: seems graph related

File: INSTALLING.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+|`PROVIDER_ACCESS_TOKEN`|no|Optional access token to be used with requests to provider graph query endpoint|_none_|
+|`PROVIDER_ACCOUNT_SEED_PHRASE`|**yes**|Seed phrase for provider control keypair|_none_|
+|`REDIS_URL`|**yes**|URL used to connect to Redis instance|_none_<br/>\*preset to the internal Redis URL in the standalone container|
+|`BLOCKCHAIN_SCAN_INTERVAL_MINUTES`|no|# of minutes to wait in between scans of the blockchain|180|
+|`QUEUE_HIGH_WATER`|no|# of pending graph scan queue entries to allow before pausing blockchain scanning until the next scan cycle|1000|
+|`GRAPH_ENVIRONMENT_TYPE`|**yes**|Indicates which blockchain network to connect to.<br/>Possible values:<br/>* `Mainnet`<br/>* `Rococo`<br/>* `Dev`|_none_|
+|`GRAPH_ENVIRONMENT_DEV_CONFIG`|no<br/>*required for 'Dev' Graph Environment|JSON configuration object for GraphSDK configuration. Used to test against a local development Frequency node|_none_|
```
Comment (aramikm) [inline INSTALLING.md:72]: graph related

File: package.json [hunk 7]
```diff
@@ -0,0 +1,86 @@
+{
+  "name": "content-publishing-service",
+  "version": "1.0.0",
```
Comment (aramikm) [inline package.json:3]: nit:  maybe `0.0.1` ?

File: package.json [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": "https://github.com/AmplicaLabs/content-publishing-service/issues"
+  },
+  "homepage": "https://github.com/AmplicaLabs/content-publishing-service#readme",
+  "dependencies": {
+    "@dsnp/graph-sdk": "^0.0.11",
```
Comment (aramikm) [inline package.json:33]: nit: do we need graph-sdk?

File: src/blockchain/blockchain-constants.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    query: string;
+  }
+
+  interface IChainRpc {
+    rpcPallet: string;
+    rpc: string;
+  }
+
+  const PALLET_FREQ_TX_PYMT = 'frequencyTxPayment';
+  const PALLET_STATEFUL_STORAGE = 'statefulStorage';
```
Comment (aramikm) [inline src/blockchain/blockchain-constants.ts:23]: should replaced with messages-pallet

File: src/blockchain/blockchain-constants.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+  const EX_PAY_CAPACITY_BATCH = 'payWithCapacityBatchAll';
+  const EX_UPSERT_PAGE = 'upsertPage';
```
Comment (aramikm) [inline src/blockchain/blockchain-constants.ts:26]: add_ipfs_message

File: src/config/config.service.spec.ts [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+
+import { Test } from '@nestjs/testing';
+import { describe, it, expect, beforeAll, jest } from '@jest/globals';
+import { ConfigModule } from '@nestjs/config';
+import { ConfigService } from './config.service';
+import { configModuleOptions } from './env.config';
+
+const setupConfigService = async (envObj: any): Promise<ConfigService> => {
+  jest.resetModules();
+  Object.keys(process.env).forEach((key) => {
```
Comment (aramikm) [inline src/config/config.service.spec.ts:14]: question: why are we removing these values?

**Review by aramikm — APPROVED** [review]
Overall looks great. Had some nits and question.

**Comment by saraswatpuneet** [comment]
> there is an assumption that we might use redis which might be correct but other than that shouldn't Redis start before the app? 

True that is something we would want to do in reconnection-service as well to prevent injection or inject only after it connects, we may want to do that in this work, but this is a just placeholder

**Comment by saraswatpuneet** [comment]
> question: why are we removing these values? 

We might not need all of them, some are related to graphsdk, we will add/remove env variables (check .env.template) and also env.config.ts

File: src/config/config.service.spec.ts (hunk 11 above)
Comment (saraswatpuneet) [reply src/config/config.service.spec.ts:14]: It just resets env variable for some different tests
spec.ts are test files

**Pull request merged**